### PR TITLE
Single component redraw fix

### PR DESCRIPTION
**Pull request by dhinesh03**
Applied @jonahx ’s single component redraw patch on the current master.

### CODE REVIEW

File: mithril.js [hunk 1]
```diff
@@ -460,10 +460,23 @@ var m = (function app(window, undefined) {
...
-	function buildObject(data, cached, editable, parentElement, index, shouldReattach, namespace, configs) {
+	function buildObject(parentElement, parentTag, parentCache, parentIndex, data, cached, shouldReattach, index, editable, namespace, configs) {
 		var views = [], controllers = [];
 		data = markViews(data, cached, views, controllers);
 		if (!data.tag && controllers.length) throw new Error("Component template must return a virtual element, not an array, string, etc.");
+		// Single component redraw fix
+		for (var i = 0; i < controllers.length; i++) {
+			var controller = controllers[i];
+			if (isFunction(controller.redrawSelf)) {
+				controller.redrawSelf = (function(ctrl,view) {
```
Comment (Naddiseo) [inline mithril.js]: https://jslinterrors.com/dont-make-functions-within-a-loop

**Comment by dhinesh03** [comment]
@Naddiseo Thanks for pointing it out.

File: mithril.js [hunk 2]
```diff
@@ -460,10 +460,18 @@ var m = (function app(window, undefined) {
...
+			var newData = view(controller);
+			build(parentElement, parentTag, parentCache, parentIndex, newData, cached, shouldReattach, index, editable, namespace, configs);							
+		}
+	}
+
+	function buildObject(parentElement, parentTag, parentCache, parentIndex, data, cached, shouldReattach, index, editable, namespace, configs) {
 		var views = [], controllers = [];
 		data = markViews(data, cached, views, controllers);
-		if (!data.tag && controllers.length) throw new Error("Component template must return a virtual element, not an array, string, etc.");
+		if (!data.tag && controllers.length) throw new Error("Component template must return a virtual element, not an array, string, etc.");
```
Comment (Naddiseo) [inline mithril.js]: trailing whitespace

File: mithril.js [hunk 3]
```diff
@@ -487,6 +495,15 @@ var m = (function app(window, undefined) {
 		//schedule configs to be called. They are called after `build`
 		//finishes running
 		scheduleConfigsToBeCalled(configs, data, node, isNew, cached);
+		
+		// Single component redraw fix
+		for (var i = 0; i < controllers.length; i++) {
+			var controller = controllers[i];
+			if (isFunction(controller.redrawSelf)) {
+				controller.redrawSelf = makeRedrawSelf(controller,views[i],parentElement, parentTag, parentCache, parentIndex, data, cached, shouldReattach, index, editable, namespace, configs);
```
Comment (Naddiseo) [inline mithril.js]: Need spaces after commas

**Comment by Naddiseo** [comment]
How does `redrawSelf` get set initially?

Is there a way to merge the for-loop with another for-loop called indirectly earlier on in the function?

**Comment by dhinesh03** [comment]
@Naddiseo 
1) We should set redrawSelf as empty function where ever needed.
 ```
redrawSelf = function () {
        };
```
Refer @jonahx 's [codepen](http://codepen.io/jonahx/pen/zGEoQX) for more details.
2) I've tried, but the cached object will be overwritten on the following line
```
cached = reconstructCached(data, attrs, children, node, namespace, views, controllers); 
```

**Comment by andre-gh** [comment]
This pull request should be considered (or similar functionality), not only is it important for optimization but also for integration with components where you know you need a full redraw such as integration with a third party library taking ownership of some DOM component after you have rendered in some initial data where you want a full redraw of that component before giving up ownership of it again.

**Comment by avesus** [comment]
@dhinesh03 please, consider adding a documentation for `controller.redrawSelf()`. Also, could we write a test for this?

**Comment by dhinesh03** [comment]
@avesus Busy with my work, Will add doc soon.

**Comment by sebastiansandqvist** [comment]
@dhinesh03 I'd definitely like to see something like this in a future release. What is the status of this pr?

**Comment by isiahmeadows** [comment]
Closing for now (cleaning up old PRs). Please open a new issue or PR if you want this to be fixed/added/implemented.

**Comment by jonahx** [comment]
Closing it and reopening it seems like it makes things messier, to me.  I'm still interested in this, and based on thread comments no reason to think other people's interest has changed.

**Comment by isiahmeadows** [comment]
You want to run a diff for yourself? It has tons of merge conflicts, and is best just redone.

**Comment by jonahx** [comment]
I’m happy to redo it if there are any plans to merge it.  If not, 
nothing has changed from the original PR, so it’s probably more 
accurate to close it as WON’T FIX.

**Comment by isiahmeadows** [comment]
If you clean this up (trust me in that you'd have to deal with a massive
merge conflict), I can reopen it once it's updated.

**Comment by willycamargo** [comment]
Any plans to implement and merge it? It is a great feature for some applications.

My application has many components working at the same time, and sometimes this feature would be good.

**Pull request closed without merging**